cmd/buildctl/build/ocilayout.go

```diff
@@ -0,0 +1,27 @@
+package build
+
+import (
+	"strings"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/content/local"
+	"github.com/pkg/errors"
+)
+
+// ParseOCILayout parses --oci-layout
```

Comment:
Could you update `README.md` to explain the usage?

Reply:
CC @deitch :eyes:

Sorry to bump this PR, but was trying to use this earlier and struggled to work out the right flags :tada:

Reply:
Don't be sorry @jedevc ; if it needs addressing, it should be raised. FWIW, I wrote this PR, and I still need to look back to recall what flags should be used where.

@AkihiroSuda I am fine opening a PR to update the README, but I am not sure where to put it. I don't see a section in the README that shows "here is where you add source contexts".

I also (despite having written this) don't remember the magical invocation. I think it was something like this. For Dockerfile:

```dockerfile
FROM akihiro/abc:123
```

```
buildctl build --opt context:foo=oci-layout:///dir/to/layout
```

but something doesn't look right to me. Something is missing about mapping `akihiro/abc:123` to context `foo`, so that `foo` can map to the oci-layout. Same thing if I were to replace it with `docker-image://library/alpine:3.16` or something else.

Help me remember how to do this, and where it goes, and I am happy to add it to the README.

Reply:
> I am not sure where to put it

Doesn't look like we have anywhere for it atm, ideally we'd have a new section on "contexts" maybe right before **Outputs**? (so we'd cover `local`, `docker-container`, `oci-layout`, `git`, etc - I'm happy to take a look at the others if you could do the `oci-layout`).

> Help me remember how to do this

No idea :smile: I'm at a bit of a loss.

I imagine something like this to work: `--opt context:foo=oci-layout://path/to/oci@sha256:6f8bc6b53cc1f29d3ae7542bf7935762ccdff29ab8866e5d3fd85f6b190748f5` - but this gives:

```
error: failed to solve: unable to get info about digest: Unimplemented: unknown service containerd.services.content.v1.Content: unknown
```

I imagine this isn't working since it doesn't look like `solveOpt.OCIStores` is even populated, since the `--oci-layout` option isn't parsed (but `--oci-layout` is also never even created as a valid option). I'm kinda confused, it seems like something is definitely broken here (I'm not sure if it ever work with just buildctl, it only seems to have been tested using buildx, which has the right oci-layout parsing logic)

Reply:
> I'm happy to take a look at the others if you could do the oci-layout).

I don't mind doing that. Why don't you get the others in, and then I can add, maybe on your PR for that?

> No idea 😄 I'm at a bit of a loss.

Haha! That makes both of us.

>  since the --oci-layout option isn't parsed (but --oci-layout is also never even created as a valid option).

I don't think it ever was; wasn't it just for `--opt` and never a full-fledged `--oci-layout`? Or maybe it was? @tonistiigi  walked me through the implementation step-by-step, as I really struggled to get my head around it.

The docker buildx extensions have a straight mapping of "image name" -> context, e.g. [here](https://www.docker.com/blog/dockerfiles-now-support-multiple-build-contexts/), so I can do:

```dockerfile
FROM akihiro/abc:123
```

and then

```
docker buildx build --build-context akihiro/abc:123=oci-layout:///path/to/layout@sha256:abcdefg11111222
```

But as I recall, the `buildctl` part was always two steps:

1. Map the image name or `AS` alias in the dockerfile to a named context, e.g. `akihiro/abc:123` --> `foo`
2. Map the named context to a source, e.g. `foo` ---> `oci-layout:///path/to/layout@sha256:abcdefg11111222`

The `--opt context:foo=...` solves the second part, but not the first. I have no memory how to do that from CLI.

> I'm not sure if it ever work with just buildctl, it only seems to have been tested using buildx, which has the right oci-layout parsing logic

I really do not recall. Tonis had us do the server-side logic first along with frontend, then buildx. I had though buildctl had the support, since it doesn't need to be explicitly listed (part 1 already works, and part 2 just gets passed to buildkitd), but maybe I am missing something?

I really thought we had run it back when this went in.

> I imagine something like this to work...  but this gives:

Can you post a sample? Here or on a gist? Let's have a common base to work with?

Reply:
To create the OCI folder:
```bash
$ rm -rf dump; echo "FROM alpine:3.14" | buildx build -f - --output type=oci,dest=./out.tar .; mkdir dump; (cd dump; tar xf ../out.tar)
```

With the following `Dockerfile`:
```dockerfile
FROM foo
RUN apk add git
```

And the following command (assuming a running buildkit built off of master listening on port 1234):

```bash
$ buildctl --addr tcp://localhost:1234 build --frontend=dockerfile.v0 --local context=. --local dockerfile=. --opt context:foo=docker-image://alpine
```

Reading through the PR, it looks like we should need to pass `--oci-layout` here, but that logic isn't plumbed through to cobra properly.

Reply:
cool, will comment there.